### ISSUE TITLE
storage: avoid key comparisons by comparing timestamps instead

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/cockroach#85610]
When iterating, we can sometimes avoid expensive key comparisons by comparing the timestamps instead: if we're still on the same key, we expect the timestamp to decrease. See also: https://github.com/cockroachdb/pebble/pull/1840.

However, there are complications here. We need to take into account intents keys (with timestamp 0) which incur an additional comparison, and it's possible for intents to cause multiple values at the same timestamp (one in the intent record, one in the provisional value).

An initial attempt with `pebbleMVCCScanner.nextKey()` and `pebbleIterator.NextKey()` taking the above into account did not show a significant improvement on `BenchmarkMVCCScan_Pebble` (in fact, it regressed). However, it still seems worth pursuing.

```
name                                                    old time/op    new time/op    delta
MVCCScan_Pebble/rows=1/versions=1/valueSize=64-24         5.46µs ± 1%    5.44µs ± 1%    ~     (p=0.118 n=10+10)
MVCCScan_Pebble/rows=1/versions=10/valueSize=64-24        7.39µs ± 1%    7.42µs ± 1%    ~     (p=0.089 n=10+10)
MVCCScan_Pebble/rows=100/versions=1/valueSize=64-24       39.4µs ± 1%    40.0µs ± 2%  +1.57%  (p=0.000 n=9+10)
MVCCScan_Pebble/rows=100/versions=10/valueSize=64-24       123µs ± 1%     123µs ± 1%    ~     (p=0.278 n=9+10)
MVCCScan_Pebble/rows=10000/versions=1/valueSize=64-24     2.87ms ± 1%    2.87ms ± 1%    ~     (p=0.968 n=9+10)
MVCCScan_Pebble/rows=10000/versions=10/valueSize=64-24    10.5ms ± 2%    10.4ms ± 2%    ~     (p=0.105 n=10+10)
```

**Comment by erikgrinaker** [comment]
>  not show a significant improvement on `BenchmarkMVCCScan_Pebble`

This is because this benchmark writes all keys at the same timestamp.

**Comment by sumeerbhola** [comment]
> it's possible for intents to cause multiple values at the same timestamp (one in the intent record, one in the provisional value)

I don't understand this -- both `pebbleIterator` and `pebbleMVCCScanner` are operating at a level where they either don't parse `MVCCMetadata` or when they do, don't peer into the `IntentHistory`.

**Comment by erikgrinaker** [comment]
Yeah, I didn't think so either, but I got a bunch of `TestMVCCHistories` test failures where we're suddenly either surfacing multiple keys with the same timestamp, or skipping these keys when we shouldn't. These were all related to txn seqnum processing. My guess is some interaction between `pebbleMVCCScanner` and `intentInterleavingIter`, but I haven't dug into it.

**Comment by erikgrinaker** [comment]
@tbg investigated this a bit over in #86546.
